server/embedded/src/org/labkey/embedded/EmbeddedExtractor.java

```diff
@@ -16,11 +22,126 @@ public class EmbeddedExtractor
 {
     private static final int BUFFER_SIZE = 1024 * 64;
 
-    private void extractExecutableJar(File jarFilePath, File destDirectory, boolean remotePipeline)
+    private final File currentDir = new File("").getAbsoluteFile();
+    private final File labkeyServerJar;
+
+    private String labkeyWebappDirName = null;
+
+    public EmbeddedExtractor()
+    {
+        File[] files = currentDir.listFiles(file -> {
+            String name = file.getName().toLowerCase();
+            return name.endsWith(".jar") && name.contains("labkeyserver");
+        });
+
+        if (files == null || files.length == 0)
+        {
+            labkeyServerJar = null;
+        }
+        else if (files.length > 1)
+        {
+            throw new ConfigException("Multiple jars found - " + Arrays.asList(files) + ". Must provide only one jar.");
+        }
+        else
+        {
+            labkeyServerJar = files[0];
+        }
+    }
+
+    public boolean foundLabkeyServerJar()
+    {
+        return labkeyServerJar != null;
+    }
+
+    private File verifyJar()
+    {
+        if (labkeyServerJar == null)
+        {
+            throw new ConfigException("Executable jar not found in " + currentDir);
+        }
+
+        return labkeyServerJar;
+    }
+
+    public boolean shouldUpgrade(File webAppLocation)
+    {
+        File existingVersionFile = new File(webAppLocation, "WEB-INF/classes/VERSION");
+
+        // Upgrade from standalone Tomcat installation
+        if (!existingVersionFile.exists())
+            return true;
+
+        String existingVersion;
+        try
+        {
+            existingVersion = Files.readString(existingVersionFile.toPath()).trim();
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException(e);
+        }
+
+        String newVersion = getNewVersion();
+
+        Version v1 = getLabKeyVersion(existingVersion);
+        Version v2 = getLabKeyVersion(newVersion);
+
+        return v1.compareTo(v2) > 0;
```

Comment:
How does this handle snapshot vs snapshot comparisons?

Additionally, I feel like we should always extract whenever the version has changed. It won't be common, but it's possible that you'd test out, say 24.3.5 and revert to 24.3.4. With this code you'd need to manually delete the extracted files first.

Reply:
I removed the complicated version check. We now redeploy if the VERSION file has changed at all.
I've updated the VERSION file to have a link to the TeamCity build to further uniquify distributions.
```
24.3-SNAPSHOT
https://teamcity.labkey.org/build/2941427
```
It will also check for changes to the `distribution` file so that it will redeploy if we push somebody a different distribution of the same release.